### ISSUE TITLE
"Lecture start" and "Lecture end" not working

### ISSUE THREAD

**Issue by tagwato** [FaisalUmair/udemy-downloader-gui#9]
Great job!

But, I have encountered some issues:

- when setting "Lecture start" and "Lecture end", then the downloaded files are given different sequence numbers (not the original order).

- when setting "Video quality" = 1080p, always got 360p. 

(Win-64, pre-compiled version from Github)

**Comment by FaisalUmair** [comment]
Hi, Thanks for reporting. 

I just tried the 64 bit windows version and it worked fine for me. I guess when you say the lectures are given different sequence numbers, you probably mean they don't start with the number "1". 

If that is the case, then let me tell you it is totally intentional and was implemented to maintain the course structure. 
So for example, if we have a course that has 2 lectures in first chapter, 4 in second and 3 in fourth. Now if your lecture start is at "4" and lecture end is at "8". The first chapter will be skipped since if only contains 2 lecture and you want your download to start from 4th lecture.   
So the 4th lecture here means 2nd lecture from chapter 2. Therefore the downloaded course folder will have its first folders listed  as "2. CHAPTER_NAME" and "3.CHAPTER_NAME" and inside the "2.CHAPTER_NAME" you will have "2.LECTURE_NAME" since the 4th lecture means 2nd lecture from 2nd Chapter and so on. 

This is done only to maintain the downloaded course structure so that it maps exactly to the course structure on Udemy. 

However, if you are already aware of this and this is not what you are referring to, then please add more details in the comments so that we get to the point.


Now coming to the video quality part, there are two possible cases where the quality you selected is not downloaded.
1. Either the download quality is not available for that lecture.
2. The course content is already fetched and the download links are already generated.  This can happen when you pause the download and resume it after changing the quality. 

However if neither of the 2 cases is true for your issue, then please explain it in more detail and add links to the Udemy course and mention the video quality you are trying to download and for what lecture.

And just to let you know, I am working on new update and from this issue one thing that I think needs to be addressed is the change of video quality after pausing the download. I think the fact that the download resumes without any warning is misleading as the user would think that the quality has been changed. I will definitely fix this either by adding a warning/notice or by updating the download links. 

Thanks.

**Comment by tagwato** [comment]
Hi, Faisal,

About de video quality, it seems you are right.

Regarding the lectures sequence numbers, I comparede the behaviour of this software with udemy-dl.exe. Please, see screenshots below. 
You said that Udeler goal is to maintain the downloaded course structure, but it seems a bit odd to me that I specify that I want to download lessons 35, 36 and 37 and what I get are lessons 5, 1, 2. 

Anyway, you are doing a great job. Thanks for this software.


[Screenshots.pdf](https://github.com/FaisalUmair/udemy-downloader-gui/files/1575543/Screenshots.pdf)

ps - it would be great if Udeler downloaded also the "attached files" of each chapter, the same way udemy-dl.exe does.

**Comment by FaisalUmair** [comment]
Thanks for adding more detail to it. Well from what I said earlier, this isn't something that should happen. Initially I thought you are referring to the case that the lecture names do not start with "1". 

Now from the screenshots I see that the sequence is incorrect. But can you confirm if the correct lectures were downloaded?. Looking at the size of the video it seems that the correct video is downloaded. but the labels are incorrect.

I will check whats wrong and fix it. And please add detail when you are reporting an issue as you can see it makes a lot of difference. I had to assume and guess what could be going on. But now I have the issue presented perfectly before me  👍

**Comment by tagwato** [comment]
Yes, the files are correct, only the labels/numbers are incorrect.

**Comment by FaisalUmair** [comment]
Alright Sir, I have added this to To Do in v1.3.0

Thanks for reporting! It really helps in improving the software.

**Comment by FaisalUmair** [comment]
Hi, While debugging I found that there isn't actually anything wrong with the sequence. I know I just admitted in one of my previous comments that the sequence is incorrect but the correct lectures are downloaded. I am sorry about that I think I got confused with the udemy-dl.exe situation. 

Let me explain this, 

This is the course we are dealing with and all the examples are based on the same course.
https://www.udemy.com/excel-2013-completo/

Now, if you notice when you are downloading with Udeler vs Udemy-dl.exe , the lecture naming scheme is not same.

Udeler prefixes the lecture number based on the current chapter where as Udemy-dl.exe prefixes the lecture number based on the whole course content. 

Here you have set that you want to download lecture 35,36,37 (35-37). If you look at the course content the 35th lecture of the course is the 5th lecture of the 6th chapter. Hence you get a folder by name 6.CHAPTER_NAME and inside it you get 5.LECTURE_NAME. Similarly for 36th lecture of course you get a folder named 7.CHAPTER_NAME and inside it you have 1.LECTURE_NAME and 2.LECTURE_NAME because 36th lecture and 37th lecture are the 1st and 2nd lectures of the 7th Chapter of the course. I hope this makes sense. 

What udemy-dl.exe does it names the lecture based on its position in the entire course and not in the chapter it belongs to. 

When you see 5.LECTURE_NAME inside 6.CHAPTER_NAME it doesn't mean that the 5th lecture of the course was downloaded it means 5th lecture of 6th chapter was downloaded which is 35th lecture of the course.

I understand it gets a bid odd when we are specifying Lecture start and End, but its just naming preference and could be changed by just changing the naming scheme from chapter based naming to course content based naming if the need arises.

Please let me know your thoughts.
